File: scripts/reconstruct-b52-d12-2-static-floor.py
#!/usr/bin/env python3
"""Scalar Python static bilinear consumer for B52-D12.2."""

from __future__ import annotations

import argparse
import hashlib
import json
import math
import os
import sys
from pathlib import Path

import numpy as np


SPEC_SHA256 = "fa63daa0c3b7b3f080a488aa0fc84996fd52cd731efce94ebe28bbc81b55d9d3"
INPUTS = {
    "previousRgba": ("previous.rgba32", 4),
    "currentRgba": ("current.rgba32", 4),
    "previousOwner": ("previous-owner.f32", 1),
    "currentOwner": ("current-owner.f32", 1),
    "vector": ("vector.xy32", 2),
}


def sha256_bytes(value: bytes) -> str:
    return hashlib.sha256(value).hexdigest()


def sha256_file(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def canonical_hash(value: object) -> str:
    return hashlib.sha256(json.dumps(value, sort_keys=True, separators=(",", ":"), allow_nan=False).encode()).hexdigest()


def arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--spec", type=Path, required=True)
    parser.add_argument("--fixture", required=True)
    parser.add_argument("--repeat", type=int, choices=(1, 2), required=True)
    parser.add_argument("--input-dir", type=Path, required=True)
    parser.add_argument("--adapter-report", type=Path, required=True)
    parser.add_argument("--output-dir", type=Path, required=True)
    parser.add_argument("--report", type=Path, required=True)
    return parser.parse_args()


def main() -> None:
    args = arguments()
    spec = json.loads(args.spec.read_text())
    if sha256_file(args.spec) != SPEC_SHA256:
        raise RuntimeError("B52-D12.2 spec identity mismatch")
    if sha256_file(Path(sys.executable)) != spec["runtime"]["python"]["sha256"] or np.__version__ != spec["runtime"]["python"]["numpy"]:
        raise RuntimeError("Python runtime identity mismatch")
    fixture = next((row for row in spec["fixtures"] if row["id"] == args.fixture), None)
    if fixture is None:
        raise RuntimeError("fixture outside D12.2 roster")
    if args.output_dir.exists() or args.report.exists():
        raise RuntimeError("refusing to overwrite D12.2 consumer output")
    adapter = json.loads(args.adapter_report.read_text())
    adapter_body = {key: value for key, value in adapter.items() if key != "reportHash"}
    if adapter.get("reportHash") != canonical_hash(adapter_body) or adapter.get("fixtureId") != args.fixture or adapter.get("repeat") != args.repeat:
        raise RuntimeError("D12.2 adapter identity mismatch")
    width, height = fixture["resolution"]
    arrays = {}
    for name, (filename, channels) in INPUTS.items():
        path = args.input_dir / filename
        payload = path.read_bytes()
        if sha256_bytes(payload) != adapter["arrays"][name]["sha256"]:
            raise RuntimeError(f"D12.2 adapter array mismatch: {name}")
        shape = (height, width, channels) if channels > 1 else (height, width)
        arrays[name] = np.frombuffer(payload, dtype="<f4").reshape(shape).copy()

    previous = arrays["previousRgba"]
    current = arrays["currentRgba"]
    reconstructed = current.copy()
    valid = np.zeros((height, width), dtype=np.uint8)
    owner_id = np.float32(fixture["passIndex"])
    margin = 4
    for y in range(margin, height - margin):
        for x in range(margin, width - margin):
            if arrays["currentOwner"][y, x] != owner_id or current[y, x, 3] <= np.float32(0.999):
                continue
            vector_x, vector_y = float(arrays["vector"][y, x, 0]), float(arrays["vector"][y, x, 1])
            qx, qy = x + vector_x, y - vector_y
            x0, y0 = math.floor(qx), math.floor(qy)
            x1, y1 = x0 + 1, y0 + 1
            if x0 < 0 or y0 < 0 or x1 >= width or y1 >= height:
                continue
            taps = ((y0, x0), (y0, x1), (y1, x0), (y1, x1))
            if not all(arrays["previousOwner"][ty, tx] == owner_id and previous[ty, tx, 3] > np.float32(0.999) for ty, tx in taps):
                continue
            fx, fy = qx - x0, qy - y0
            w0, w1, w2, w3 = (1.0 - fx) * (1.0 - fy), fx * (1.0 - fy), (1.0 - fx) * fy, fx * fy
            for channel in range(4):
                v0, v1 = float(previous[y0, x0, channel]), float(previous[y0, x1, channel])
                v2, v3 = float(previous[y1, x0, channel]), float(previous[y1, x1, channel])
                reconstructed[y, x, channel] = np.float32((((v0 * w0) + (v1 * w1)) + (v2 * w2)) + (v3 * w3))
            valid[y, x] = 1

    args.output_dir.mkdir(parents=True, exist_ok=False)
    records = {}
    for name, filename, array, dtype in (
        ("reconstructed", "reconstructed.rgba32", reconstructed, "little-endian-float32"),
        ("valid", "valid.u8", valid, "uint8"),
    ):
        payload = np.ascontiguousarray(array, dtype="<f4" if name == "reconstructed" else "u1").tobytes(order="C")
        target = args.output_dir / filename
        target.write_bytes(payload)
        records[name] = {"uri": str(target), "sha256": sha256_bytes(payload), "bytes": len(payload), "shape": list(array.shape), "dtype": dtype}
    report = {
        "schemaVersion": "bfs.blenderStaticVectorFloorConsumerReport.v0.1",
        "experimentId": spec["experimentId"],
        "producer": "python",
        "fixtureId": args.fixture,
        "repeat": args.repeat,
        "pid": os.getpid(),
        "runtime": {"python": sys.version.split()[0], "pythonExecutableSha256": sha256_file(Path(sys.executable)), "numpy": np.__version__},
        "adapter": {"uri": str(args.adapter_report), "sha256": sha256_file(args.adapter_report), "reportHash": adapter["reportHash"]},
        "contract": spec["consumer"],
        "arrays": records,
        "integrity": "external dual typed-envelope sidecars",
        "operationCounts": {"consumerProcesses": 1, "pixelsVisited": width * height, "modelCalls": 0, "networkCalls": 0},
    }
    args.report.parent.mkdir(parents=True, exist_ok=True)
    args.report.write_text(json.dumps(report, indent=2, sort_keys=True, allow_nan=False) + "\n")
    print(f"BFS_B52_D122_CONSUMER_PY_OK fixture={args.fixture} repeat={args.repeat} valid={int(valid.sum())}")


if __name__ == "__main__":
    main()

File: scripts/reconstruct-b52-d12-2-static-floor.mjs
#!/usr/bin/env node
/** Scalar Node static bilinear consumer for B52-D12.2. */

import crypto from 'node:crypto';
import fs from 'node:fs';
import process from 'node:process';

const SPEC_SHA256 = 'fa63daa0c3b7b3f080a488aa0fc84996fd52cd731efce94ebe28bbc81b55d9d3';
const INPUTS = {
  previousRgba: ['previous.rgba32', 4], currentRgba: ['current.rgba32', 4],
  previousOwner: ['previous-owner.f32', 1], currentOwner: ['current-owner.f32', 1], vector: ['vector.xy32', 2],
};

function parseArgs() {
  const result = {};
  for (let index = 2; index < process.argv.length; index += 2) result[process.argv[index].replace(/^--/, '')] = process.argv[index + 1];
  for (const key of ['spec', 'fixture', 'repeat', 'input-dir', 'adapter-report', 'output-dir', 'report']) if (!(key in result)) throw new Error(`missing --${key}`);
  result.repeat = Number(result.repeat);
  if (![1, 2].includes(result.repeat)) throw new Error('repeat outside D12.2 roster');
  return result;
}

function shaBytes(value) { return crypto.createHash('sha256').update(value).digest('hex'); }
function shaFile(path) { return shaBytes(fs.readFileSync(path)); }
function stable(value) {
  if (Array.isArray(value)) return `[${value.map(stable).join(',')}]`;
  if (value && typeof value === 'object') return `{${Object.keys(value).sort().map(key => `${JSON.stringify(key)}:${stable(value[key])}`).join(',')}}`;
  return JSON.stringify(value);
}
function canonicalHash(value) { return shaBytes(Buffer.from(stable(value))); }
function readF32(path) {
  const buffer = fs.readFileSync(path), values = new Float32Array(buffer.length / 4);
  for (let index = 0; index < values.length; index++) values[index] = buffer.readFloatLE(index * 4);
  return { buffer, values };
}
function encodeF32(values) {
  const buffer = Buffer.alloc(values.length * 4);
  for (let index = 0; index < values.length; index++) buffer.writeFloatLE(values[index], index * 4);
  return buffer;
}
function pixelBase(width, x, y, channels) { return (y * width + x) * channels; }

function main() {
  const args = parseArgs(), specBuffer = fs.readFileSync(args.spec);
  if (shaBytes(specBuffer) !== SPEC_SHA256) throw new Error('B52-D12.2 spec identity mismatch');
  const spec = JSON.parse(specBuffer);
  if (process.version !== spec.runtime.node.version || shaFile(process.execPath) !== spec.runtime.node.sha256) throw new Error('Node runtime identity mismatch');
  const fixture = spec.fixtures.find(row => row.id === args.fixture);
  if (!fixture) throw new Error('fixture outside D12.2 roster');
  if (fs.existsSync(args['output-dir']) || fs.existsSync(args.report)) throw new Error('refusing to overwrite D12.2 consumer output');
  const adapter = JSON.parse(fs.readFileSync(args['adapter-report'], 'utf8'));
  const adapterBody = Object.fromEntries(Object.entries(adapter).filter(([key]) => key !== 'reportHash'));
  if (adapter.reportHash !== canonicalHash(adapterBody) || adapter.fixtureId !== args.fixture || adapter.repeat !== args.repeat) throw new Error('D12.2 adapter identity mismatch');
  const [width, height] = fixture.resolution, arrays = {};
  for (const [name, [filename]] of Object.entries(INPUTS)) {
    const loaded = readF32(`${args['input-dir']}/${filename}`);
    if (shaBytes(loaded.buffer) !== adapter.arrays[name].sha256) throw new Error(`D12.2 adapter array mismatch: ${name}`);
    arrays[name] = loaded.values;
  }
  const previous = arrays.previousRgba, current = arrays.currentRgba;
  const reconstructed = new Float32Array(current), valid = new Uint8Array(width * height);
  const ownerId = Math.fround(fixture.passIndex), margin = 4;
  for (let y = margin; y < height - margin; y++) for (let x = margin; x < width - margin; x++) {
    const pixel = y * width + x, currentBase = pixel * 4;
    if (arrays.currentOwner[pixel] !== ownerId || current[currentBase + 3] <= Math.fround(0.999)) continue;
    const vectorX = arrays.vector[pixel * 2], vectorY = arrays.vector[pixel * 2 + 1];
    const qx = x + vectorX, qy = y - vectorY, x0 = Math.floor(qx), y0 = Math.floor(qy), x1 = x0 + 1, y1 = y0 + 1;
    if (x0 < 0 || y0 < 0 || x1 >= width || y1 >= height) continue;
    const taps = [[y0, x0], [y0, x1], [y1, x0], [y1, x1]];
    if (!taps.every(([ty, tx]) => {
      const tap = ty * width + tx;
      return arrays.previousOwner[tap] === ownerId && previous[tap * 4 + 3] > Math.fround(0.999);
    })) continue;
    const fx = qx - x0, fy = qy - y0;
    const w0 = (1 - fx) * (1 - fy), w1 = fx * (1 - fy), w2 = (1 - fx) * fy, w3 = fx * fy;
    for (let channel = 0; channel < 4; channel++) {
      const v0 = previous[pixelBase(width, x0, y0, 4) + channel], v1 = previous[pixelBase(width, x1, y0, 4) + channel];
      const v2 = previous[pixelBase(width, x0, y1, 4) + channel], v3 = previous[pixelBase(width, x1, y1, 4) + channel];
      reconstructed[currentBase + channel] = Math.fround((((v0 * w0) + (v1 * w1)) + (v2 * w2)) + (v3 * w3));
    }
    valid[pixel] = 1;
  }
  fs.mkdirSync(args['output-dir'], { recursive: true });
  const reconstructedPayload = encodeF32(reconstructed), validPayload = Buffer.from(valid);
  const reconstructedPath = `${args['output-dir']}/reconstructed.rgba32`, validPath = `${args['output-dir']}/valid.u8`;
  fs.writeFileSync(reconstructedPath, reconstructedPayload); fs.writeFileSync(validPath, validPayload);
  const report = {
    schemaVersion: 'bfs.blenderStaticVectorFloorConsumerReport.v0.1', experimentId: spec.experimentId, producer: 'node', fixtureId: args.fixture,
    repeat: args.repeat, pid: process.pid, runtime: { node: process.version, nodeExecutableSha256: shaFile(process.execPath) },
    adapter: { uri: args['adapter-report'], sha256: shaFile(args['adapter-report']), reportHash: adapter.reportHash }, contract: spec.consumer,
    arrays: {
      reconstructed: { uri: reconstructedPath, sha256: shaBytes(reconstructedPayload), bytes: reconstructedPayload.length, shape: [height, width, 4], dtype: 'little-endian-float32' },
      valid: { uri: validPath, sha256: shaBytes(validPayload), bytes: validPayload.length, shape: [height, width], dtype: 'uint8' },
    },
    integrity: 'external dual typed-envelope sidecars', operationCounts: { consumerProcesses: 1, pixelsVisited: width * height, modelCalls: 0, networkCalls: 0 },
  };
  fs.mkdirSync(args.report.slice(0, args.report.lastIndexOf('/')), { recursive: true });
  fs.writeFileSync(args.report, `${JSON.stringify(report, null, 2)}\n`);
  process.stdout.write(`BFS_B52_D122_CONSUMER_NODE_OK fixture=${args.fixture} repeat=${args.repeat} valid=${valid.reduce((a, b) => a + b, 0)}\n`);
}

main();
